### PR TITLE
[Improvement] Test with onnx models and TensorRT engines.

### LINKED ISSUE #614: Inference after converting from PyTorch to Onnx to TensorRT
I've converted the PyTorch TSN model to Onnx and I'm planning to convert it to TensorRT.  I wanted to know how can I make a prediction for a video input using the TensorRT model. Do I need to make any changes in the `Prediction.py` file ?

### PR DESCRIPTION
**Pull request by irvingzhang0512**
Fix #614 

## Description
+ Run `tools/test.py` with onnx models and TesnorRT engines instead of pytorch models.
+ For now, only support fixed shape and single gpu

## Usage
+ export onnx models by `tools/pytorch2onnx.py`.
+ generate tensorrt engines by official tensorrt tools.
+ Make sure the input tensors of onnx/tensor models and the output tensors of dataset share the same shape.
  + onnx/tensorrt model input shape is set by `[--shape ${SHAPE}]` in `tools/pytorch2onnx.py`.
  + The shape of dataset is set by `videos_per_gpu`(batch size) and test pipelines(all kinds of crop/resize, and num_clips for dense sampling, test augmentation like `ThreeCrop/TenCrop/twice_sample`, etc. )
  + It is recommended to remove all test augmentations. 
    + The input shapes of onnx/tensorrt models should be `1, num_segments, 3, height, width` for 2D Recognizers and `1, 1, 3, num_segments, height, width` for 3D recognizers.
    + set `videos_per_gpu=1` and remove test augmentation pipelines(`ThreeCrop/TenCrop/twice_sample`, etc.)

```python
data = dict(
    videos_per_gpu=8,  # default value for train/val/test
    workers_per_gpu=4,
    test_dataloader=dict(videos_per_gpu=1), # specific value for test
    train=dict(...),
    val=dict(...),
    test=dict(...),
)
```

+ Run test script
```
python tools/test.py /path/to/config.py /path/to/model.onnx --onnx --out test.json

python tools/test.py /path/to/config.py /path/to/model.trt --tensorrt --out test.json
```

## TODO
+ [x] codes in test.py
+ [x] docs in getting started.
+ test all kinds of models
  + [x] tsm
  + [x] i3d
  + [x] tsn
  + [x] slowonly
  + [x] slowfast

### CODE REVIEW

**Comment by irvingzhang0512** [comment]
Any comments about this pr? @innerlee 

I have no idea how to add unittest about onnx/tensorrt inference.

**Comment by dreamerlin** [comment]
cc. @RunningLeon

**Comment by RunningLeon** [comment]
@dreamerlin Currently, we only add unit tests for plugins of TensorRT.

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
@congee524 Could you have a try? Will merge this if you can run it successfully :)

**Comment by congee524** [comment]
OKK

**Comment by congee524** [comment]
Hey, I have tested the script and passed.
Since we don't support doing `average_clip` after inferring the class scores in ONNX models, I think it's necessary to add notes in the docstring to remind users.
i.e. don't use `ThreeCrops`, `TenCrops`, `Twice_sample` etc. to test ONNX models that were transformed with our `tools/pytorch2onnx.py`.

**Comment by irvingzhang0512** [comment]
1. For the `clip_average` issue, I think we should add docs for `pytorch2onnx.py`, not add docs for this pr.

Maybe we should add an option to export onnx with softmax. I'll look into it.

2. I'll add some docs for the `ThreeCrop/TenCrop` issue.

File: docs/getting_started.md [hunk 1]
```diff
@@ -86,8 +86,8 @@ Optional arguments:
...
-- `--tensorrt`: If specified, recognition results will be generated by TensorRT engine and `CHECKPOINT_FILE` should be TensorRT engine file path. TensorRT engines files are generated by exported onnx models and TensorRT official convertion tools. For now, only support single-gpu mode and fixed input shape. Please note that dataset and TensorRT engine should share the same batch size.
+- `--onnx`: If specified, recognition results will be generated by onnx model and `CHECKPOINT_FILE` should be onnx model file path. Onnx model files are generated by `/tools/pytorch2onnx.py`. For now, multi-gpu mode and dynamic input shape mode are not supported. Please note that the output tensors of dataset and the input tensors of onnx model should share the same shape. And it is recommended to remove all test augmentation pipelines(`ThreeCrop`, `TenCrop`, `twice_sample`, etc.)
```
Comment (congee524) [inline docs/getting_started.md:89]: And it is recommended to remove all data augmentation methods in the `test_pipeline` (`ThreeCrop`, `TenCrop`, `twice_sample`, etc.)

File: docs/getting_started.md [hunk 2]
```diff
@@ -86,8 +86,8 @@ Optional arguments:
...
+- `--tensorrt`: If specified, recognition results will be generated by TensorRT engine and `CHECKPOINT_FILE` should be TensorRT engine file path. TensorRT engines are generated by exported onnx models and TensorRT official convertion tools. For now, multi-gpu mode and dynamic input shape mode are not supported. Please note that the output tensors of dataset and the input tensors of TensorRT engine should share the same shape. And it is recommended to remove all test augmentation pipelines(`ThreeCrop`, `TenCrop`, `twice_sample`, etc.)
```
Comment (congee524) [inline docs/getting_started.md:90]: same

File: docs/getting_started.md (hunk 1 above)
Comment (irvingzhang0512) [reply docs/getting_started.md:89]: Since `test time augmentation` is a widely used term, how about "test time augmentation methods in `test_pipeline`"

Comment (congee524) [reply docs/getting_started.md:89]: testing augmentation?

Comment (congee524) [reply docs/getting_started.md:89]: test-time augmentation?

Comment (irvingzhang0512) [reply docs/getting_started.md:89]: You're right

**Review by congee524 — APPROVED** [review]

**Comment by congee524** [comment]
Happy April Fool's Day :p

**Comment by innerlee** [comment]
Thanks!
